### PR TITLE
Add option to disable TCP tunneling by filter state

### PR DESCRIPTION
**Pull request by ohadvano**
Commit Message: option to disable TCP tunneling per connection, by filter state
Additional Description: Adding an option to disable TCP tunneling by setting a filter state object for the key `envoy.tcp_proxy.disable_tunneling`. If this filter state object is set, then tunneling will not take place, even if configured in the filter configuration.
Risk Level: low
Testing: unit tests
Docs Changes: tcp_proxy configuration reference
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto
Comment (kyessenov) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:205]: _disable tunneling per connection_

File: envoy/stream_info/boolean_accessor.h
Comment (kyessenov) [inline envoy/stream_info/boolean_accessor.h:17]: I think in the past we used a string accessor with "reason" as value. That might be more informative than a boolean.

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (kyessenov) [inline source/extensions/upstreams/tcp/generic/config.cc:22]: Need a constant defined I think

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (kyessenov) [inline source/extensions/upstreams/tcp/generic/config.cc:19]: Could we have a test in tcp_proxy for this feature?

**Review by kyessenov — COMMENTED** [review]
Left initial comments.

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto
Comment (ohadvano) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto]: done

File: envoy/stream_info/boolean_accessor.h
Comment (ohadvano) [reply envoy/stream_info/boolean_accessor.h]: I thought about it being confusing to users, but I guess that it works if it's a pattern already in use. I deleted the boolean accessor and used string accessor instead.

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (ohadvano) [reply source/extensions/upstreams/tcp/generic/config.cc]: done

Comment (ohadvano) [reply source/extensions/upstreams/tcp/generic/config.cc]: @kyessenov I added a constant, but in tcp_proxy.h. When I come to think about it, the tcp_proxy implementation is receiving a generic upstream from the generic pool factory (which is an extension, in extensions/upstreams/tcp/generic/config.cc) and the logic to inspect the existence of that filter state key, is at the factory implementation. Do you think I should move the constant definition there and rename?

On second hand, if it's defined in the extensions, generally it won't be accessible to external consumers due to visibility, although in this particular case the "config" target is defined as public.

 WDYT? Please help me decide on this.

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (ohadvano) [reply source/extensions/upstreams/tcp/generic/config.cc:19]: tcp_proxy uses a generic upstream created by the generic connection pool factory, so this feature was added to the factory implementation. I added tests to the generic pool factory, LMK if you see any other required test case, IMO this should be sufficient.

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (kyessenov) [reply source/extensions/upstreams/tcp/generic/config.cc]: I think it should be defined at the point of use, which is tcp_proxy in common code. That way it's shared and accessible everywhere.

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (kyessenov) [inline source/extensions/upstreams/tcp/generic/config.cc:20]: This block should be moved under `config.has_value()` to avoid lookup for non-tunnel TCP streams.

**Review by kyessenov — COMMENTED** [review]
One minor performance improvement, but otherwise LGTM. Please follow the PR template.

File: source/extensions/upstreams/tcp/generic/config.cc
Comment (ohadvano) [reply source/extensions/upstreams/tcp/generic/config.cc]: done

**Review by kyessenov — DISMISSED** [review]

**Comment by ohadvano** [comment]
@adisuissa can you please have a look at this PR?

**Comment by adisuissa** [comment]
> @adisuissa the use case I encountered is based on internal business logic that basically checks some pre-conditions in a network filter that is running prior to tcp_proxy. On most cases that logic would result in continuing as usual, and the traffic would be routed through tcp_proxy and tunneling is applied. In some specific edge cases, there's a requirement to send the traffic to a different kind of upstream, and there tunneling is not required (and can't be used...). In that case, I use the "[dynamic cluster selection](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/tcp_proxy_filter#dynamic-cluster-selection)" feature (that also uses filter state to override config) to route to a different upstream, and I would like to use the feature I introduced here to disable tunneling.
> 
> I agree that this adds another complexity to the system, but I encountered this practice of overriding configs dynamically through filter state in multiple occasions.

Thanks for "dynamic cluster selection" link. I agree this is not the first use-case, but I wonder if this could be done in a better way.
I think it would have been better if there was a network composite filter (similar to the [http composite filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/composite_filter)), that will use the [filter-state matcher](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/filter_state.proto#type-matcher-v3-filterstatematcher) to decide which tcp-proxy filter will be used.
@envoyproxy/senior-maintainers for thoughts on dynamic override of tcp-proxy configuration using filter-state values.

**Comment by ohadvano** [comment]
@adisuissa that's a good thinking but I'm not sure it would be trivial to make this because there needs to be a terminal filter for a filter chain, which is either tcp_proxy or http_connection_manager. For your suggestion it would require supporting another terminal filter

/assign-from @envoyproxy/senior-maintainers

**Comment by yanavlasov** [comment]
I do not think we have a mechanism that can be used to achieve desired effect with the tcp_proxy, since there is really no routing of TCP connections. Hence the dynamic cluster selection. 

I think short of implementing routing for TCP connections, this is the only approach.

**Review by yanavlasov — DISMISSED** [review]

**Comment by yanavlasov** [comment]
Asking for non Google maintainer to review as well.

File: source/common/tcp_proxy/upstream.h
Comment (ggreenway) [inline source/common/tcp_proxy/upstream.h:18]: To match other uses of FilterState, I think it would be better to define a specific filter state object that should be set, instead of any object with the given name, similar to `class PerConnectionCluster : public StreamInfo::FilterState::Object` (from tcp_proxy.h), and give it a boolean value.

**Review by ggreenway — CHANGES_REQUESTED** [review]
I agree with @yanavlasov that this approach is the best available at this time.

/wait

File: source/common/tcp_proxy/upstream.h
Comment (ohadvano) [reply source/common/tcp_proxy/upstream.h:18]: @ggreenway Let me quote Lizan from [another PR I merged some time ago](https://github.com/envoyproxy/envoy/pull/25234), Lizan asked to use StringAccessor, quoting his reasoning:

> "You should still use the key name envoy.upstream.dynamic_host but the type of filter object should be a StringAccessor and I don't see why it needs to be a custom type. A string accessor is easier to get/set by other filters and for [logging and header substitution](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#config-access-log-format-filter-state)"

And quoting Kuat from this PR:

> I think in the past we used a string accessor with "reason" as value. That might be more informative than a boolean.

I personally I don't have a hard preference between these two options, they both makes sense to me, but I tend towards StringAccessor. What should we agree about?

Comment (kyessenov) [reply source/common/tcp_proxy/upstream.h:18]: @ggreenway IMHO the type system for filter states is based on their key name, not the C++ type. We don't have any "reflection" yet based on the content, so it's all just a blob for consumption sites.

Comment (ggreenway) [reply source/common/tcp_proxy/upstream.h:18]: @kyessenov I don't think that's true; most/all FilterState has specific types. Each key has a specific type, so that something like `connection.streamInfo().filterState()->getDataReadOnly<PerConnectionCluster>()` will work.

I'm fine with using a common type for that (maybe add `BoolAccessor`), but I think putting the value into an Object, instead of just depending on whether the key exists, is clearer.

Comment (ohadvano) [reply source/common/tcp_proxy/upstream.h:18]: done

**Review by ggreenway — APPROVED** [review]
LGTM. Thanks!

**Comment by adisuissa** [comment]
@ravenblackx @kyessenov and I discussed this.
It would be great if the tcp-filter will have an explicit config knob that allows enabling the new filter-state setting, as well as having a metric that tracks whenever that setting is modifying the filter's behavior.
This will aid in debugging/monitoring because:
1. The config knob will allow someone looking at the config be aware that it is could impact the filter's behavior.
2. The metric tells the user that the filter behavior was indeed altered in runtime.

Ideally there will be a mechanism for any cross-filter-setting using a filter-state, and the filter will "register" these settings.

I don't feel strongly about this at the moment, because as you've mentioned there are other cases where this is being used. @ravenblackx and @kyessenov to chime if they think differently.

/lgtm api

**Comment by ohadvano** [comment]
@adisuissa considering that this option doesn't kick in unless explictly enabled by filter state (or any other filter state controlled overrides), and considering that enabling this by filter state is necessarily involving a piece of custom code that sets this filter state object, I don't see why a user couldn't know what kind of behavior to expect at runtime as you can log in case the filter state was set to override in your code. I didn't fully get the config knob idea. As I see it  the filter state itself is the config knob, the dev/user needs to be aware of this feature to enable it in first place. Please let me know if I missed something.

**Comment by ggreenway** [comment]
I agree with @ohadvano. Having a required config knob is just another hoop the user has to jump through to make the feature work, and if they forget to set it, it's completely unclear why things are not working as expected. I don't like the idea of adding a config knob for this.

**Comment by kyessenov** [comment]
I don't think we should block this PR because we don't have mechanisms in place to solve the data dependency issues in general. We shouldn't be doing one-off metric or access control in `tcp_proxy`,  it should be a general control over which filter states are allowed to propagate and whether they have an effect. I'd imagine this would be a filter state API next to `getData` that says data was used.

**Comment by adisuissa** [comment]
> I agree with @ohadvano. Having a required config knob is just another hoop the user has to jump through to make the feature work, and if they forget to set it, it's completely unclear why things are not working as expected. I don't like the idea of adding a config knob for this.

@ggreenway we do get some questions from time to time on trying to debug why is the filter behaving one way or another. As it stands, there is no way to understand this cross-filter impact because there isn't any stat that shows it, and there isn't a config knob that clarifies that this is the way it works.
Another example is when using different versions of Envoy (or filters) and these kind of changes could impact behavior and go unnoticed. The feature isn't even described in the [docs[(https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/tcp_proxy_filter).

From a high-level point of view, setting cross-filter fields should be done with care and be observable. As it stands, there is no way to know which filter may change the behavior of the Tcp-Proxy filter, other than searching the code of all filters.

That said, I don't feel strongly that we should change this PR (other than suggesting updating the docs of the filter).

**Comment by ggreenway** [comment]
@adisuissa I agree it would be very useful to have an operator be able to see the current mode of operation of tcp_proxy (or any other filter). I just don't like the idea of requiring a user to manually set a knob to do so; it's something we know internally needs to be set (via filterstate) so requiring a manual action from the user seems unfriendly.

Let's brainstorm in some other forum (slack, issue, etc) the best way to do that.

**Comment by adisuissa** [comment]
> @adisuissa I agree it would be very useful to have an operator be able to see the current mode of operation of tcp_proxy (or any other filter). I just don't like the idea of requiring a user to manually set a knob to do so; it's something we know internally needs to be set (via filterstate) so requiring a manual action from the user seems unfriendly.
> 
> Let's brainstorm in some other forum (slack, issue, etc) the best way to do that.

SGTM. Adding you to the Slack thread where we discussed this with @ravenblackx and @kyessenov.

**Review by ggreenway — APPROVED** [review]
Thanks!

**Pull request merged**
